### ISSUE TITLE
Panic when removing task, what volume removal fails

### ISSUE THREAD

**Issue by jeichorn** [aws/amazon-ecs-agent#2058]
### Summary
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xc05a48]
```

### Environment Details
```
Containers: 154
 Running: 33
 Paused: 0
 Stopped: 121
Images: 26
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.14.114-105.126.amzn2.x86_64
Operating System: Amazon Linux 2
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.446GiB
Name: ip-10-99-14-118.ec2.internal
ID: 2RW4:CIYS:QRYT:MBTH:6CZW:57EF:D2BX:E2FP:PNAM:V6U2:6T4O:ZNII
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

```
{"Cluster":"XXX","ContainerInstanceArn":"arn:aws:ecs:us-east-1:XXX:container-instance/d38eddd5-779e-4af9-8d8b-7104445ba1b1","Version":"Amazon ECS Agent - v1.28.0 (d2456beb)"}
```

### Supporting Log Snippets

```
2019-05-31T17:14:47Z [INFO] Removing Container Reference: mgmt-worker-cdn-init from Image State- sha256:6a12ec016215c60a3e7322889294f8937392770362a00d9f603346dcd3b9a52f
2019-05-31T17:14:47Z [INFO] Saving state! module="statemanager"
2019-05-31T17:14:47Z [INFO] Task engine [arn:aws:ecs:us-east-1:XXX:task/bb2d8aa7-2acd-4430-8ac2-4a27dcbe3abd]: removing container: mgmt-worker-cdn-worker
2019-05-31T17:14:47Z [INFO] Removing Container Reference: mgmt-worker-cdn-worker from Image State- sha256:5509a443eea4ea49f1a2124bbd0c5db9e618b6c989fb6fe906791e32656742b4
2019-05-31T17:14:47Z [INFO] Saving state! module="statemanager"
2019-05-31T17:14:48Z [INFO] Task engine [arn:aws:ecs:us-east-1:XXX:task/bb2d8aa7-2acd-4430-8ac2-4a27dcbe3abd]: removing container: mgmt-worker-cdn-sidecar
2019-05-31T17:14:48Z [INFO] Removing Container Reference: mgmt-worker-cdn-sidecar from Image State- sha256:9875aad13e4708afb56827762055e95211a46149b40dcfcfcf6371ccd558a13a
2019-05-31T17:14:48Z [INFO] Saving state! module="statemanager"
2019-05-31T17:14:48Z [INFO] Saving state! module="statemanager"
2019-05-31T17:14:48Z [DEBUG] Removing cgroup /ecs/bb2d8aa7-2acd-4430-8ac2-4a27dcbe3abd
2019-05-31T17:14:48Z [INFO] Task engine [arn:aws:ecs:us-east-1:XXX:task/bb2d8aa7-2acd-4430-8ac2-4a27dcbe3abd]: resource cgroup cleanup complete
2019-05-31T17:14:48Z [DEBUG] Removing volume with name app
2019-05-31T17:14:49Z [INFO] Task engine [arn:aws:ecs:us-east-1:XXX:task/bb2d8aa7-2acd-4430-8ac2-4a27dcbe3abd]: resource app cleanup complete
2019-05-31T17:14:49Z [DEBUG] Removing volume with name sidecar
2019-05-31T17:14:49Z [INFO] Task engine [arn:aws:ecs:us-east-1:XXX:task/bb2d8aa7-2acd-4430-8ac2-4a27dcbe3abd]: resource sidecar cleanup complete
2019-05-31T17:14:49Z [DEBUG] Removing volume with name logs
2019-05-31T17:14:49Z [INFO] Task engine [arn:aws:ecs:us-east-1:XXX:task/bb2d8aa7-2acd-4430-8ac2-4a27dcbe3abd]: resource logs cleanup complete
2019-05-31T17:14:49Z [DEBUG] Removing volume with name app
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0xc05a48]

goroutine 3177 [running]:
github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi.CannotRemoveVolumeError.Error(...)
        /go/src/github.com/aws/amazon-ecs-agent/agent/dockerclient/dockerapi/errors.go:308
github.com/aws/amazon-ecs-agent/agent/taskresource/volume.(*VolumeResource).Cleanup(0xc00333dd10, 0x2e, 0xc0032fc180)
        /go/src/github.com/aws/amazon-ecs-agent/agent/taskresource/volume/dockervolume.go:330 +0x1e9
github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).deleteTask(0xc000154a20, 0xc003372540)
        /go/src/github.com/aws/amazon-ecs-agent/agent/engine/docker_task_engine.go:487 +0x96
github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).cleanupTask(0xc000c888f0, 0x9d29229e000)
        /go/src/github.com/aws/amazon-ecs-agent/agent/engine/task_manager.go:1128 +0x2b1
github.com/aws/amazon-ecs-agent/agent/engine.(*managedTask).overseeTask(0xc000c888f0)
        /go/src/github.com/aws/amazon-ecs-agent/agent/engine/task_manager.go:250 +0x1f1
created by github.com/aws/amazon-ecs-agent/agent/engine.(*DockerTaskEngine).startTask
        /go/src/github.com/aws/amazon-ecs-agent/agent/engine/docker_task_engine.go:543 +0x94
```

**Comment by yunhee-l** [comment]
Thank you for reporting this issue. 
We are looking into fixing it, but it would help us a great deal if we can reproduce this - do you mind sending the task definition you used and possibly docker log when this happened please? you can send them to yhlee _at_ amazon _dot_ com.

**Comment by yunhee-l** [comment]
Thank you for the logs and task definitions; we have identified a code bug and fixed it here: https://github.com/aws/amazon-ecs-agent/pull/2059  This will be included in our next release, so in the mean time please continue using the workaround.

**Comment by daleevans** [comment]
> Thank you for the logs and task definitions; we have identified a code bug and fixed it here: #2059 This will be included in our next release, so in the mean time please continue using the workaround.

We are experiencing this issue as well, but I don't see which workaround you are talking about?

**Comment by jeichorn** [comment]
Our workaround has been to add retries to fetching metadata, the aws sdk's are pretty flexible so its not that hard to implement, just annoying since the agent can take a long time to restart on a busy system.

**Comment by fierlion** [comment]
https://github.com/aws/amazon-ecs-agent/pull/2059 has been released as a part of the latest agent v1.28.1 which fixed the issue
